dbms/src/Storages/MergeTree/DiskSpaceMonitor.h

```diff
@@ -140,6 +154,62 @@ class DiskSpaceMonitor
         return std::make_unique<Reservation>(size);
     }
 
+    /// Returns mount point of filesystem where absoulte_path (must exist) is located
+    static std::filesystem::path getMountPoint(std::filesystem::path absolute_path)
+    {
+        if (absolute_path.is_relative())
+            throw Exception("Path is relative. It's a bug.", ErrorCodes::LOGICAL_ERROR);
+
+        absolute_path = std::filesystem::canonical(absolute_path);
+
+        const auto get_device_id = [](const std::filesystem::path & p)
+        {
+            struct stat st;
+            if (stat(p.c_str(), &st))
+                throwFromErrno("Cannot stat " + p.string(), p.string(), ErrorCodes::SYSTEM_ERROR);
+            return st.st_dev;
+        };
+
+        /// If /some/path/to/dir/ and /some/path/to/ have different device id,
+        /// then device which contains /some/path/to/dir/filename is mounted to /some/path/to/dir/
+        auto device_id = get_device_id(absolute_path);
+        while (absolute_path.has_relative_path())
+        {
+            auto parent = absolute_path.parent_path();
+            auto parent_device_id = get_device_id(parent);
+            if (device_id != parent_device_id)
+                return absolute_path;
+            absolute_path = parent;
+            device_id = parent_device_id;
+        }
+
+        return absolute_path;
+    }
+
+    /// Returns name of filesystem mounted to mount_point
+#if !defined(__linux__)
+[[noreturn]]
+#endif
+    static std::string getFilesystemName([[maybe_unused]] const std::string & mount_point)
+    {
+#if defined(__linux__)
+        auto mounted_filesystems = setmntent("/etc/mtab", "r");
```

Comment:
Just in case, we need to check "the internets", is it a good function (never heard about it).

Reply:
It seems to me, there is nothing suspicious:
- `setmntent` opens `/etc/mtab` (or `/etc/fstab`) using `fopen`
- `getmntent_r` parses opened file line-by-line
- `endmntent` calls `fclose`
 
However, `findmnt` uses libmount and `df` uses `read_file_system_list()` from gnulib to parse `/proc/self/mountinfo` instead of `/etc/mtab`.